### DISCUSSION TITLE
Why is the thrown exception not printed in the background thread?

### DISCUSSION THREAD

**Discussion by a13782425** [timcassell/ProtoPromise#448] [Q&A] [answered]
Why is the thrown exception not printed in the background thread?
Code 
```csharp
private async void RunTest()
{
    Debug.LogError("RunTest:1");
    Debug.LogError("ThreadId:"+Thread.CurrentThread.ManagedThreadId);
    await PromiseYielder.WaitForRealTime(2);
    await Promise.SwitchToBackgroundAwait();
    Debug.LogError("ThreadId:"+Thread.CurrentThread.ManagedThreadId);
    await PromiseYielder.WaitForRealTime(2);
    Debug.LogError("cccc");
    await Promise.SwitchToForegroundAwait();
    Debug.LogError(Thread.CurrentThread.ManagedThreadId);
    Debug.LogError("ttt");
    await PromiseYielder.WaitForRealTime(2);
    Debug.LogError("RunTest:2");
}
```
![image](https://github.com/timcassell/ProtoPromise/assets/11312875/ab13d2d6-5e05-4832-893b-3848ebeb1a67)

**Comment by timcassell** [comment]
Probably because you're using `async void` which is using the built-in Tasks system, not ProtoPromise. I'm not entirely sure how it handles uncaught exceptions. Try refactoring to this:

```cs
private void RunTest()
{
    Run().Forget();
    
    async Promise Run()
    {
        Debug.LogError("RunTest:1");
        Debug.LogError("ThreadId:"+Thread.CurrentThread.ManagedThreadId);
        await PromiseYielder.WaitForRealTime(2);
        await Promise.SwitchToBackgroundAwait();
        Debug.LogError("ThreadId:"+Thread.CurrentThread.ManagedThreadId);
        await PromiseYielder.WaitForRealTime(2);
        Debug.LogError("cccc");
        await Promise.SwitchToForegroundAwait();
        Debug.LogError(Thread.CurrentThread.ManagedThreadId);
        Debug.LogError("ttt");
        await PromiseYielder.WaitForRealTime(2);
        Debug.LogError("RunTest:2");
    }
}
```

**Reply by a13782425** [reply]
Thanks, this does get exceptions, why can't get exceptions using `async void` which is using the built-in Tasks system?

**Reply by timcassell** [reply, marked as answer]
I'm guessing it has to do with the issue described in this forum thread: https://forum.unity.com/threads/suppressing-not-logging-exceptions-from-threadpool-and-async-void-methods-from-non-main-thread.1059491/

**Reply by a13782425** [reply]
This is a difficult question.

**Reply by timcassell** [reply]
See also https://github.com/timcassell/ProtoPromise/issues/137#issuecomment-1341904544

**Reply by a13782425** [reply]
Except for `Run().Forget();`, Does it have a blocking version? Similar to` Run().GetAwaiter().GetResult(); `, but calling it throws an exception

**Reply by timcassell** [reply]
You can use `.Wait()`. Be careful, though, as you may deadlock if the operation tries to switch to the same thread.

**Reply by a13782425** [reply]
Thank you for your patience
